File: intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if headA is None or headB is None: #List is empty
            return None
        temp = headA
        lenA = 0
        while temp: #Find length of list A
            temp = temp.next
            lenA += 1
        temp = headB
        lenB = 0
        while temp: #Find length of list B
            lenB += 1
            temp = temp.next
        while lenA > lenB or lenB > lenA: #Shorten the longer list, intersection must exist in shorter list if it at all
            if lenA > lenB:
                headA = headA.next
                lenA -= 1
            else:
                headB = headB.next
                lenB -= 1
        while lenA > 0: #Look for intersection
            if headA == headB:
                return headA
            else:
                headA = headA.next
                headB = headB.next
                lenA -= 1
        return None

File: intersection-of-two-linked-lists/intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution 
{
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) 
    {
        // One list does not exist, no intersection
        if (headA == null)
            return null;
        if (headB == null)
            return null;
        
        // Find the length of each list
        int lengthA = 0, lengthB = 0;
        ListNode itA = headA, itB = headB;
        while (itA != null)
        {
            lengthA++;
            itA = itA.next;
        }
        while (itB != null)
        {
            lengthB++;
            itB = itB.next;
        }
        
        // Find the equivalent start point
        itA = headA;
        itB = headB;
        while (lengthA > lengthB)
        {
            itA = itA.next;
            lengthA--;
        }
        while (lengthB > lengthA)
        {
            itB = itB.next;
            lengthB--;
        }
        
        // Find the possible intersection
        while (lengthA > 0 && lengthB > 0)
        {
            if (itA == itB)
                return itA;
            itA = itA.next;
            itB = itB.next;
            lengthA--;
            lengthB--;
        }
        return null;
    }
}